src/category_theory/presheaf.lean

```diff
@@ -0,0 +1,161 @@
+-- Copyright (c) 2018 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Scott Morrison, Mario Carneiro, Reid Barton
+
+import category_theory.instances.topological_spaces
+import category_theory.functor_category
+import category_theory.whiskering
+import category_theory.natural_isomorphism
+import category_theory.opposites
+import topology.basic
+
+open topological_space
+
+universes v u
+
+@[simp] lemma eq.trans_refl_left {α : Sort u} {p q : α} (h : p = q) : eq.trans (eq.refl _) h = h := rfl
+@[simp] lemma eq.trans_refl_right {α : Sort u} {p q : α} (h : p = q) : eq.trans h (eq.refl _) = h := rfl
+
+@[simp] lemma congr_arg_refl {α : Sort u} {β : Sort v} {f : α → β} {a : α} :
+  congr_arg f (eq.refl a) = eq.refl (f a) := rfl
+
+@[simp] lemma congr_refl_fun {α : Sort u} {β : Sort v} {f : α → β} {a₁ a₂ : α} (h : a₁ = a₂) :
+  congr (eq.refl f) h = congr_arg f h := rfl
+
+@[simp] lemma congr_refl_arg {α : Sort u} {β : Sort v} {f₁ f₂ : α → β} {a : α} (h : f₁ = f₂) :
+  congr h (eq.refl a) = congr_fun h a := rfl
+
+open category_theory
+open category_theory.instances
+
+namespace category_theory.presheaves
+
+variables (C : Type u) [𝒞 : category.{v} C]
+include 𝒞
+
+structure Presheaf :=
```

Comment:
A term of `X : Presheaf C` is coerced to the topological space. This causes some semantic dissonance in my head. I feel like I need to train my mind to read `X : Space_with_Presheaf_of C` instead.

Reply:
perhaps a better name is in order...?

Reply:
`Space_with_Presheaf_of` is actually the best name I can think of... or maybe drop the `of`?

Reply:
I've gone with `PresheafedSpace` for now, by (loose) analogy with "ringed spaces", etc.